static/extensions/Ikelene/email2FA.js

```diff
@@ -0,0 +1,317 @@
+            if (cachedProjectID) return cachedProjectID;
+
+            try {
+                const res = await fetch('https://ikelene.ca/api/2fa/register_project.php', {
+                    method: 'POST',
+                    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+                    body: `project=${encodeURIComponent(args.PROJECTNAME)}`
+                });
+
+                const id = await res.text();
+                const trimmed = id.trim();
+
+                const isError = ['invalid name', 'invalid chars', 'rate limit', 'rate limited'].includes(trimmed);
+                if (trimmed && !isError) {
+                    cachedProjectID = trimmed;
+
+                    const runtime = Scratch.runtime;
+                    if (runtime && runtime.vm) {
+                        runtime.vm.extensionState['IkeEmailTwofa'] ||= {};
+                        runtime.vm.extensionState['IkeEmailTwofa']['projectID'] = trimmed;
+                    }
+
+                    return cachedProjectID;
+                }
+
+                return '';
+            } catch (e) {
+                console.error('Failed to register project:', e);
+                return '';
+            }
+        }
+
+        async send2FACode(args) {
+            try {
+                const body = {
+                    email: args.EMAIL,
+                    project_id: args.PROJECT_ID,
+                    preset: emailPreset
+                };
+
+                if (customHTML && customEmailSettings) {
+                    body.custom_html = customHTML;
+                }
+
+                if (customEmailSettings) {
+                    body.custom_email = customEmailSettings.email;
+                    body.custom_pass = customEmailSettings.pass;
+                    body.custom_smtp = customEmailSettings.smtp;
+                    body.custom_port = customEmailSettings.port;
+                }
+
+                const response = await fetch('https://ikelene.ca/api/2fa/send_code.php', {
+                    method: 'POST',
+                    headers: { 'Content-Type': 'application/json' },
+                    body: JSON.stringify(body)
+                });
+                const result = await response.text();
+                if (!response.ok) throw new Error(result);
+                console.log("2FA send success:", result);
+            } catch (e) {
+                console.error("2FA send error:", e);
+                throw e;
+            }
+        }
+
+        async verify2FACode(args) {
+            try {
+                const res = await fetch('https://ikelene.ca/api/2fa/verify_code.php', {
+                    method: 'POST',
+                    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+                    body: new URLSearchParams({
+                        email: args.EMAIL,
+                        code: args.CODE,
+                        project_id: args.PROJECT_ID
```

Comment:
For, like, the third time, could this be gotten from the project ID stored in the cache rather than being taken in as an argument?

Reply:
Pretty much just replace any `args.PROJECT_ID` with `cachedProjectID` in your codebase, and remove all `PROJECT_ID` arguments from the `getInfo` method.